sdk/src/assertions/actions.rs

```diff
@@ -445,14 +491,67 @@ impl Action {
 
     /// Adds an ingredient id to the action.
     pub fn add_ingredient_id(&mut self, ingredient_id: &str) -> Result<&mut Self> {
-        if let Some(Value::Array(ids)) = self.get_parameter_mut(CAI_INGREDIENT_IDS) {
+        if let Some(Value::Array(ids)) = self.get_parameter_mut("ingredientIds") {
             ids.push(Value::Text(ingredient_id.to_string()));
             return Ok(self);
         }
         let ids = vec![Value::Text(ingredient_id.to_string())];
-        self.set_parameter_ref(CAI_INGREDIENT_IDS, ids)?;
+        self.set_parameter_ref("ingredientIds", ids)?;
         Ok(self)
     }
+
+    /// Extracts ingredient IDs from the action, prioritizing ingredientIds, then org.cai.ingredientIds, then instanceId.
+    /// This is used to map actions to their associated ingredients.
+    /// We don't want any of these fields in the final CBOR, so we remove them after extracting.
+    pub(crate) fn extract_ingredient_ids(&mut self) -> Option<Vec<String>> {
+        let ingredient_ids = self.remove_parameter("ingredientIds");
+        let cai_ingredient_ids = self.remove_parameter("org.cai.ingredientIds");
+        #[allow(deprecated)]
+        let instance_id = self.instance_id.take();
+
+        let mut ids: Vec<String> = Vec::new();
+
+        let mut convert_ids = |val: Option<serde_cbor::Value>| {
+            if let Some(val) = val {
+                match val {
+                    serde_cbor::Value::Array(arr) => {
+                        for v in arr {
+                            if let serde_cbor::Value::Text(s) = v {
+                                ids.push(s);
+                            }
+                        }
+                    }
+                    serde_cbor::Value::Text(s) => ids.push(s),
+                    _ => {}
+                }
+            }
+        };
+
+        convert_ids(ingredient_ids);
+        convert_ids(cai_ingredient_ids);
+
+        if !ids.is_empty() {
+            Some(ids)
+        } else if let Some(s) = instance_id {
+            Some(vec![s])
+        } else {
+            None
+        }
+    }
+}
+
+// Converts a JSON string to an [`Action`].
+impl TryFrom<&str> for Action {
+    type Error = crate::error::Error;
+
+    fn try_from(action_json: &str) -> Result<Self> {
+        let action: Action = serde_json::from_str(action_json)?;
+
+        if V2_DEPRECATED_ACTIONS.contains(&action.action.as_str()) {
+            error!("Action '{}' is deprecated in C2PA v2", action.action);
+        }
+        Ok(action)
+    }
 }
```

Comment:
Hm, I think this would be more clear if we had a `from_json` method. Or possibly even better, we can add a `#[serde(deserialize_with = no_deprecated_actions])` attribute to the `action.action` field.

Reply:
I just deleted it since I don't really need it yet.